File: src/main/java/me/ruslanys/vkmusic/config/FxmlConfiguration.java
package me.ruslanys.vkmusic.config;

import javafx.fxml.FXMLLoader;
import me.ruslanys.vkmusic.annotation.FxmlController;
import org.reflections.Reflections;
import org.springframework.beans.BeansException;
import org.springframework.beans.factory.config.AutowireCapableBeanFactory;
import org.springframework.beans.factory.config.ConfigurableListableBeanFactory;
import org.springframework.beans.factory.support.BeanDefinitionRegistry;
import org.springframework.beans.factory.support.BeanDefinitionRegistryPostProcessor;
import org.springframework.context.ApplicationListener;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.context.event.ContextRefreshedEvent;

import java.io.IOException;
import java.io.InputStream;
import java.util.Map;
import java.util.Set;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * @author Ruslan Molchanov (ruslanys@gmail.com)
 * @author http://mruslan.com
 */
@Configuration
public class FxmlConfiguration implements ApplicationListener<ContextRefreshedEvent> {

    @Bean
    public static BeanDefinitionRegistryPostProcessor beanDefinitionRegistryPostProcessor() {
        return new RegistryPostProcessor();
    }

    public static class RegistryPostProcessor implements BeanDefinitionRegistryPostProcessor {

        @Override
        public void postProcessBeanDefinitionRegistry(BeanDefinitionRegistry registry) throws BeansException {
        }

        @Override
        public void postProcessBeanFactory(ConfigurableListableBeanFactory beanFactory) throws BeansException {
            Reflections reflections = new Reflections("me.ruslanys.vkmusic");
            Set<Class<?>> fxmlControllers = reflections.getTypesAnnotatedWith(FxmlController.class);
            Pattern pattern = Pattern.compile("(\\w+)\\.fxml$");

            for (Class<?> fxmlController : fxmlControllers) {
                FxmlController annotation = fxmlController.getAnnotation(FxmlController.class);
                String viewPath = annotation.view();

                Matcher matcher = pattern.matcher(viewPath);
                if (!matcher.find()) {
                    throw new RuntimeException("Incorrect view path.");
                }
                String prefix = matcher.group(1);

                try (InputStream fxmlStream = getClass().getClassLoader().getResourceAsStream(viewPath)) {
                    FXMLLoader loader = new FXMLLoader();
                    loader.load(fxmlStream);

                    String viewName = prefix + "View";
                    String controllerName = prefix + "FxmlController";

                    Object view = loader.getRoot();
                    Object controller = loader.getController();

                    beanFactory.registerSingleton(viewName, view);
                    beanFactory.registerSingleton(controllerName, controller);
                } catch (IOException e) {
                    throw new RuntimeException(e);
                }
            }

        }

    }

    public void onApplicationEvent(ContextRefreshedEvent event) {
        Map<String, Object> beans = event.getApplicationContext().getBeansWithAnnotation(FxmlController.class);
        AutowireCapableBeanFactory factory = event.getApplicationContext().getAutowireCapableBeanFactory();

        for (Map.Entry<String, Object> entry : beans.entrySet()) {
            factory.autowireBean(entry.getValue());
            factory.initializeBean(entry.getValue(), entry.getKey());
        }
    }

}

File: src/main/kotlin/me/ruslanys/vkmusic/config/FxmlConfiguration.kt
package me.ruslanys.vkmusic.config

import javafx.fxml.FXMLLoader
import me.ruslanys.vkmusic.annotation.FxmlController
import me.ruslanys.vkmusic.controller.BaseController
import org.reflections.Reflections
import org.springframework.beans.factory.config.ConfigurableListableBeanFactory
import org.springframework.beans.factory.support.BeanDefinitionRegistry
import org.springframework.beans.factory.support.BeanDefinitionRegistryPostProcessor
import org.springframework.context.ApplicationListener
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.context.event.ContextRefreshedEvent
import java.util.regex.Pattern

/**
 * This approach lets initialize controllers by the FXML loader.
 * And then, generates BeansDefinition at Spring context as well as registering and injecting created beans.
 *
 * So, the order of initialization is:
 * 1. Creating controllers by FXML loader.
 * 2. Registering BeansDefinitions.
 * 3. Initializing Spring Context.
 * 4. Injecting controllers dependencies.
 */
@Deprecated("This approach is complex and worse in the order of beans initialization")
class FxmlFirstConfiguration : ApplicationListener<ContextRefreshedEvent> {

    class RegistryPostProcessor : BeanDefinitionRegistryPostProcessor {

        override fun postProcessBeanDefinitionRegistry(registry: BeanDefinitionRegistry) {
            // No needs to override this behavior
        }

        override fun postProcessBeanFactory(beanFactory: ConfigurableListableBeanFactory) {
            val reflections = Reflections("me.ruslanys.vkmusic")
            val fxmlControllers = reflections.getTypesAnnotatedWith(FxmlController::class.java)
            val pattern = Pattern.compile("(\\w+)\\.fxml$")

            for (fxmlController in fxmlControllers) {
                val annotation = fxmlController.getAnnotation(FxmlController::class.java)
                val viewPath = annotation.view

                val matcher = pattern.matcher(viewPath)
                if (!matcher.find()) {
                    throw RuntimeException("Incorrect view path.")
                }
                val prefix = matcher.group(1)

                javaClass.classLoader.getResourceAsStream(viewPath).use { fxmlStream ->
                    val loader = FXMLLoader()
                    loader.load<Any>(fxmlStream)

                    val viewName = prefix + "View"
                    val controllerName = prefix + "FxmlController"

                    val view = loader.getRoot<Any>()
                    val controller = loader.getController<Any>()

                    beanFactory.registerSingleton(viewName, view)
                    beanFactory.registerSingleton(controllerName, controller)
                }
            }

        }

    }

    override fun onApplicationEvent(event: ContextRefreshedEvent) {
        val beans = event.applicationContext.getBeansWithAnnotation(FxmlController::class.java)
        val factory = event.applicationContext.autowireCapableBeanFactory

        for ((key, value) in beans) {
            factory.autowireBean(value)
            factory.initializeBean(value, key)
        }
    }

    companion object {
        @Bean
        @JvmStatic
        fun beanDefinitionRegistryPostProcessor(): BeanDefinitionRegistryPostProcessor {
            return RegistryPostProcessor()
        }
    }

}

/**
 * This approach lets initialize controllers by the Spring IoC.
 * And then, initialize them with FXML context, when Spring Context is ready.
 *
 * This approach is recommended by me :)
 *
 * Limitations:
 * 1. Do not setup fx:controller field at FXML markup.
 */
@Configuration
class FxmlConfiguration : ApplicationListener<ContextRefreshedEvent> {
    override fun onApplicationEvent(event: ContextRefreshedEvent) {
        val beans = event.applicationContext.getBeansWithAnnotation(FxmlController::class.java)
        val pattern = Pattern.compile("(\\w+)\\.fxml$")

        for (controller in beans.values) {
            val annotation = controller::class.java.getAnnotation(FxmlController::class.java)
            val viewPath = annotation.view

            val matcher = pattern.matcher(viewPath)
            if (!matcher.find()) {
                throw RuntimeException("Incorrect view path.")
            }

            javaClass.classLoader.getResourceAsStream(viewPath).use { fxmlStream ->
                val loader = FXMLLoader()
                loader.setController(controller)
                loader.load<Any>(fxmlStream)

                if (controller is BaseController) {
                    controller.rootView = loader.getRoot()
                }
            }
        }
    }

}